marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/impl/hub-entities.xqy

```diff
@@ -808,25 +837,38 @@ declare %private function hent:add-indexes-for-entity-properties($entities as js
       let $_ :=
         for $entity-type-property in map:keys($entity-type-properties)
           let $ref := map:get($entity-type-properties, $entity-type-property)=>map:get("$ref")
-          where fn:empty($ref)
           return
+            let $items := map:get($entity-type-properties, $entity-type-property)=>map:get("items")
             let $is-facetable :=
-              let $items := map:get($entity-type-properties, $entity-type-property)=>map:get("items")
-              return
-                if(fn:empty($items)) then
-                  map:get($entity-type-properties, $entity-type-property)=>map:get("facetable")
-                else
-                  fn:not(fn:starts-with(map:get($items, "$ref"), "#")) and
-                  map:get($entity-type-properties, $entity-type-property)=>map:get("facetable")
+              if (fn:empty($ref) and (fn:empty($items) or fn:not(fn:starts-with(map:get($items, "$ref"), "#")))) then
+                map:get($entity-type-properties, $entity-type-property)=>map:get("facetable")
+              else
+                let $definiton-name :=
+                  if (fn:empty($items)) then
+                    fn:substring-after($ref, "#/definitions/")
+                  else
+                    fn:substring-after(map:get($items, "$ref"), "#/definitions/")
+                let $_ :=
+                  if (fn:not(fn:not($definiton-name))) then
+                    let $definitions := map:get($model, "definitions")
+                    let $entity-path := $entity-type-property || "/" || $definiton-name
+                    let $model := hent:add-indexes-for-structured-entity-properties($entity-path, $definiton-name, $definitions, $entity-type)
```

Comment:
This is confusing - it implies that something useful is being returned that's assigned to "$model", but then "$model" is ignored. But based on the function def, "$entity-type" is what indexes are being added to.

I think the main confusion is over the name "entity-type". That implies that it's a string, but it's not - it's a map, it's the actual entity definition. So that should be renamed to "$entity-definition" (or just "$entity-def", as "def" is an intuitive abbreviation for "$definition").

"$model" should only refer to what a model is - i.e. the entire model descriptor, which contains "definitions" and "info".